src/MongoDB.Driver.Core/Core/Authentication/ScramShaAuthenticator.cs

```diff
@@ -189,22 +193,72 @@ private class ClientFirst : ISaslStep
                 _hi = hi;
                 _hmac = hmac;
                 _rPrefix = rPrefix;
+                _baseThreadId = Thread.CurrentThread.ManagedThreadId;
+                _baseCulture = CultureInfo.CurrentCulture.ToString();
             }
 
             public byte[] BytesToSendToServer => _bytesToSendToServer;
 
             public bool IsComplete => false;
 
+            string BuildErrorMessage(
+                string baseMessage,
+                string connectionId,
+                byte[] bytesToSendToServer,
+                byte[] bytesReceivedFromServer,
+                string serverFirstMessage,
+                string rPrefix,
+                string r,
+                string originalThreadId,
+                string currentThreadId,
+                string originalCulture,
+                string currentCulture,
+                string clientFirstMessageBare)
+            {
+                var builder = new StringBuilder();
+
+                builder.AppendLine("Base message:" + baseMessage);
+                builder.AppendLine($"{nameof(connectionId)}: {connectionId}");
+                builder.AppendLine($"{nameof(bytesToSendToServer)}: {string.Join(",",bytesToSendToServer)}");
```

Comment:
Have you tested this?

I'm wondering how this works when the second argument is `IEnumerable<byte>`.

Reply:
It'll be easier to deal with if you hex-encode all the byte arrays.

Reply:
@jyemin done. Now, the message will look like:

    Base message:Server sent an invalid nonce.
    connectionId: { ServerId : { ClusterId : 3, EndPoint : "Unspecified/localhost:27018" }, LocalValue : 7 }
    Hex_bytesToSendToServer: 6e2c2c6e3d7573657241646d696e342c723d655c21416c73384453503c24242a36457d5f3133
    Hex_bytesReceivedFromServer: 723d655c21416c73384453503c24242a36457d5f31336b2b34457578796a3742696346364c6d2b412b516561497075734848535538782c733d345a79735447386336345a67632f4a724b4b416e5a413d3d2c693d3130303030
    rPrefix: e\!Als8DSP<$$*6E}_13
    r: e\!Als8DSP<$$*6E}_13k+4Euxyj7BicF6Lm+A+QeaIpusHHSU8x
    serverFirstMessage: r=e\!Als8DSP<$$*6E}_13k+4Euxyj7BicF6Lm+A+QeaIpusHHSU8x,s=4ZysTG8c64Zgc/JrKKAnZA==,i=10000
    originalThreadId: 16
    currentThreadId: 17
    originalCulture: en-US
    currentCulture: en-US
    clientFirstMessageBare: n=userAdmin4,r=e\!Als8DSP<$$*6E}_13
    bytesToSendToServer: 110,44,44,110,61,117,115,101,114,65,100,109,105,110,52,44,114,61,101,92,33,65,108,115,56,68,83,80,60,36,36,42,54,69,125,95,49,51
    bytesReceivedFromServer: 114,61,101,92,33,65,108,115,56,68,83,80,60,36,36,42,54,69,125,95,49,51,107,43,52,69,117,120,121,106,55,66,105,99,70,54,76,109,43,65,43,81,101,97,73,112,117,115,72,72,83,85,56,120,44,115,61,52,90,121,115,84,71,56,99,54,52,90,103,99,47,74,114,75,75,65,110,90,65,61,61,44,105,61,49,48,48,48,48

Reply:
Looks like you have a mix of hex and CSV integers.